cmd/controller/main.go

```diff
@@ -134,9 +134,12 @@ func main() {
 			brokerInformer,
 		),
 	}
-	if len(controllers) != numControllers {
-		logger.Fatalf("Number of controllers and QPS settings mismatch: %d != %d", len(controllers), numControllers)
-	}
+	// This line asserts at compile time that the length of controllers is equal to numControllers.
+	// It is based on https://go101.org/article/tips.html#assert-at-compile-time, which notes that
+	// var _ [N-M]int
+	// asserts at compile time that N >= M, which we can use to establish equality of N and M:
+	// (N >= M) && (M >= N) => (N == M)
+	var _ [numControllers - len(controllers)][len(controllers) - numControllers]int
```

Comment:
<sing with the voice of Macklemore and tune Thrift Shop>: This is freaking awesome! <well, slighlty changed>

Reply:
I'm wary of relying on something that needs a 5-line comment to explain a 1-liner technique for compile-ish time-ish bounds checking. I don't really understand how this works and I expect that like most folks, I will _assume_ that it works and not notice when if it somehow doesn't.

I'd be more comfortable with runtime checking supported by a test (since Go's type system doesn't idiomatically encode these kinds of restrictions).

Reply:
>I'm wary of relying on something that needs a 5-line comment
I started with a one-liner comment, but was quickly (and correctly) prompted that an explanation is in order.

>I don't really understand how this works and I expect that like most folks, I will assume that it works and not notice when if it somehow doesn't.
There's no need to assume, I wouldn't wont to:
https://golang.org/ref/spec#Array_types
The length is part of the array's type; it must evaluate to a non-negative constant representable by a value of type int.

var _ [N-M]int
(N - M >=0) <=> (N => M) <=> ((N > M) || (N == M))

var _ [M-N]int
(M - N >=0) <=> (M => N) <=> ((M > N) || (M == N))

var _ [N-M][M-N]int
For what N and M
((N > M) || (N == M)) && ((M > N) || (M == N)) == true?

For integers M and N, there're only 3 ordering possibilities:
1. N > M. Substitute:
(true || false) && (false || false) == false

2. M > N
(false || false) && (true || false) == false

3. N == M
The only alternative left. But still substitute:
(false || true) && (false || true) == true

>compile-ish time-ish
I'm not familiar with that program translation phase. At the present this is a compile-time check. Even if somehow in some Go implementation far, far away that line is evaluated at runtime, it will still fail and we're going to be no worse off than we started.

The earlier a problem is detected, the quicker it can be fixed. It'd be much more productive to find a way to make that assert unnecessary in the first place.